Source/Atomic/UI/UIFinderWindow.cpp

```diff
@@ -0,0 +1,618 @@
+    
+    if ( filesystem->DirExists (currentPath ) )
+    { 
+        Vector <String> mydirs;
+        int nn = 0;
+        filesystem->ScanDir (mydirs,currentPath, "*", SCAN_DIRS, false );
+        Sort(mydirs.Begin(), mydirs.End(), CompareStrs); // sort them
+        for ( nn=0; nn<mydirs.Size(); nn++ ) 
+        {
+            if ( mydirs[nn] == "." ) continue;  
+            if ( mydirs[nn] == ".." ) continue;
+            String idz = "DIR"+ String(nn);
+            fileSource->AddItem( new UISelectItem(context_, mydirs[nn], idz, "FolderIcon" ));
+        }
+        Vector <String> myfiles;
+        filesystem->ScanDir (myfiles, currentPath, "*", SCAN_FILES, false );
+        Sort(myfiles.Begin(), myfiles.End(), CompareStrs);
+        for ( nn=0; nn<myfiles.Size(); nn++ ) 
+        {
+            String idz = "FIL"+ String(nn);
+            fileSource->AddItem(new UISelectItem( context_, myfiles[nn], idz));
+        }
+    }
+    filelist->SetValue(-1);
+    filelist->SetSource(fileSource);
+}
+
+// utility to add a folder in current path
+void UIFinderWindow::CreateFolder( const String& string )
+{
+    FileSystem* filesystem = GetSubsystem<FileSystem>();
+    if ( filesystem->CreateDir( currentPath + string ) )
+    {
+        UpdateUiList();
+    }
+}
+
+// utility to add a bookmark from the current path
+void UIFinderWindow::CreateBookmark ( const String& bkname, const String&  bkpath ) 
+{
+    UISelectList *bklist = static_cast<UISelectList *>(GetBookmarksWidget());
+    if (bklist ) 
+    {
+        int inspos = bklist->GetNumItems();
+        if ( inspos < 0 ) inspos = 0;
+        String idz = "BKM" + String(inspos);
+        if ( bklist->AddItem ( inspos, bkname, idz ) )
+        {
+            bookmarks.Push(bkname);
+            bookmarkPaths.Push (bkpath);
+            bookmarksDirty = 1;
+        }
+    }
+}
+
+// removes a bookmark based upon its index in the bookmark list
+void UIFinderWindow::DeleteBookmark ( int bkindex )
+{
+    UISelectList *bklist = static_cast<UISelectList *>(GetBookmarksWidget());
+    if (bklist && bkindex >= 0) 
+    {
+        bklist->DeleteItem(bkindex);
+        bookmarks.Erase(bkindex,1);
+        bookmarkPaths.Erase(bkindex,1);
+        resultPath = "";
+        SetCurrentPath (bookmarkPaths[0]);
+        UpdateUiPath();
+        UpdateUiList();
+        UpdateUiResult();
+        bklist->SetValue(-1);
+        bookmarksDirty = 1;
+    }
+}
+
+void UIFinderWindow::LoadBookmarks()
+{
+    FileSystem* filesystem = GetSubsystem<FileSystem>();
+    String bkdata = "";
+    String bkpath = "";
+    
+#if defined(ANDROID) || defined( __ANDROID__) || defined(IOS)
+    bkpath = filesystem->GetUserDocumentsDir(); // somewhere writable on mobile
+#else
+    bkpath = filesystem->GetAppPreferencesDir("AtomicGameEngine", "Bookmarks"); // desktop systems
+#endif
+    
+    bkpath += "/Bookmarks.txt";
+    if ( filesystem->FileExists ( bkpath ) )
```

Comment:
My choices were text, xml, json. In c++ coding the json stuff is just as complex as the xml, so I opted for lower complexity.